worker/codalabworker/worker.py

```diff
@@ -188,15 +209,60 @@ def _checkin(self):
                     self._exiting = True
                 self._should_upgrade = True
 
+    def _allocate_cpu_and_gpu_sets(self, request_cpus, request_gpus):
+        """
+        Allocate a cpuset and gpuset to assign to a bundle based on given requested resources.
+        Side effects: updates the free sets, self._cpuset_free and self._gpuset_free
+
+        Arguments:
+            request_cpus: integer
+            request_gpus: integer
+
+        Returns a 2-tuple:
+            cpuset: Allocated cpuset. Empty set if allocation was unsuccessful
+            gpuset: Allocated gpuset. Empty set if allocation was unsuccessful
+        """
+        cpuset, gpuset = set(), set()
+
+        with self._resource_lock:
+            if len(self._cpuset_free) < request_cpus or len(self._gpuset_free) < request_gpus:
+                return cpuset, gpuset
+
+            for i in range(request_cpus):
+                cpuset.add(self._cpuset_free.pop())
+            for j in range(request_gpus):
+                gpuset.add(self._gpuset_free.pop())
+            return cpuset, gpuset
+
+    def _deallocate_cpu_and_sets(self, cpuset, gpuset):
+        """
+        Release held up cpus and gpus
+
+        Re-add cpuset and gpuset back to their respective free sets
+        """
+        with self._resource_lock:
+            self._cpuset_free |= cpuset
+            self._gpuset_free |= gpuset
+
     def _run(self, bundle, resources):
         if self.shared_file_system:
             bundle_path = bundle['location']
         else:
             bundle_path = self._dependency_manager.get_run_path(bundle['uuid'])
+
+        cpuset, gpuset = self._allocate_cpu_and_gpu_sets(
+                resources['request_cpus'], resources['request_gpus'])
+
+        if len(cpuset) == 0 and len(gpuset) == 0: # revert self._cpuset_free and self._gpuset_free in-place
+            logger.debug('Unsuccessful allocation of cpu and gpu sets for bundle %s', bundle['uuid'])
```

Comment:
Likely unhandled case: what is/should be the bundle status at this point? Does it get failed automagically? Probably not.

Add a call to the server to fail the bundle here? <- is this good design?
Add a call to the server to restage the bundle here? <- is this good design?

Thoughts? @percyliang @bkgoksel

Reply:
Why would this happen?  Seems like we should work hard to make sure this doesn't happen, in which case, it doesn't matter what we do here, and the simplest thing is to just fail the bundle.